Add support for environment variables

turekj opened the pull request:
Adds support for environment variables in Xcode schemes. Example usage: snapshot tests (see: [iOSSnapshotTestCase](https://github.com/uber/ios-snapshot-test-case).

This is a base needed to implement [XcodeGen's #236](https://github.com/yonaskolb/XcodeGen/issues/236).

I have seen that integration tests are failing, but going through the diff I think it was not caused by this PR. Also, it improves Eigen integration test case.

Let me know if it's any good for you! 👍

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/xcodeswift/xcproj/227)
<!-- Reviewable:end -->

yonaskolb commented:
Nice work @turekj!
Have you check that the written order of xml elements matches what xcode writes? We want to keep the work achieved by #216. You may have to add something to the attributesOrder in AEXML+XcodeFormat.swift

turekj commented:
Thanks for the great tip @yonaskolb! 💯 

I have double checked that and it seems that the order matches perfectly 😨. But, since it's a dictionary passed around, it's probably due to a sheer luck. I have updated `attributesOrder` so that it contains an entry for `EnvironmentVariable` too. 👍

turekj commented:
One more thing, I have just checked and it seems that `MacroExpansion` option should precede both `CommandLineArguments` and `EnvironmentVariables` which is not the case for `XCScheme` currently.

My Xcode's output:

```xml
<LaunchAction
   buildConfiguration = "Debug"
   selectedDebuggerIdentifier = "Xcode.DebuggerFoundation.Debugger.LLDB"
   selectedLauncherIdentifier = "Xcode.DebuggerFoundation.Launcher.LLDB"
   language = "en"
   launchStyle = "0"
   useCustomWorkingDirectory = "NO"
   ignoresPersistentStateOnLaunch = "NO"
   debugDocumentVersioning = "YES"
   debugServiceExtension = "internal"
   allowLocationSimulation = "YES">
   <MacroExpansion>
      <BuildableReference
         BuildableIdentifier = "primary"
         BlueprintIdentifier = "8C8CACE81F68184F00EFB5A7"
         BuildableName = "ProjectUnitTests.xctest"
         BlueprintName = "ProjectUnitTests"
         ReferencedContainer = "container:Project.xcodeproj">
      </BuildableReference>
   </MacroExpansion>
   <CommandLineArguments>
      <CommandLineArgument
         argument = "LAUNCH_ARGUMENT"
         isEnabled = "YES">
      </CommandLineArgument>
   </CommandLineArguments>
   <EnvironmentVariables>
      <EnvironmentVariable
         key = "FB_REFERENCE_IMAGE_DIR"
         value = "$(SOURCE_ROOT)/Snapshots/ReferenceImages"
         isEnabled = "YES">
      </EnvironmentVariable>
      <EnvironmentVariable
         key = "IMAGE_DIFF_DIR"
         value = "$(SOURCE_ROOT)/Snapshots/FailureDiffs"
         isEnabled = "YES">
      </EnvironmentVariable>
   </EnvironmentVariables>
   <AdditionalOptions>
      <AdditionalOption
         key = "NSZombieEnabled"
         value = "YES"
         isEnabled = "YES">
      </AdditionalOption>
   </AdditionalOptions>
</LaunchAction>
```

I will update the code to reflect the order accordingly. This can be always reverted 👍

yonaskolb commented on Sources/xcproj/XCScheme.swift:
```diff
@@ -169,6 +169,56 @@ final public class XCScheme {
         }
     }
 
+    final public class EnvironmentVariables {
```
Not sure if we need this class, as it only simply holds an array of `[EnvironmentVariable]`.

Maybe we can factor this and `CommandLineArguments` away later

turekj commented on Sources/xcproj/XCScheme.swift:
```diff
@@ -169,6 +169,56 @@ final public class XCScheme {
 
+    final public class EnvironmentVariables {
+        public struct EnvironmentVariable {
+            public let key: String
```
I thought about that and finally decided against it to be consistent with `xcproj` naming. 

However, since you mention that too I will refactor it and see how it goes 👍

turekj commented:
@yonaskolb I have renamed key to variable and also removed the `EnvironmentVariables` container. I haven't touched `CommandLineArguments` because that would be a breaking change.

Let me know if there's anything more I can do 👍

pepibumur commented:
Excellent job @turekj. The first implementation of xcproj was done with immutable structs but at some point we turned them into classes. We find it a bit cumbersome because we needed a lot of API methods to create copies of the objects mutating some attributes. There would be a benefit on using immutable models here if we were doing multi-threading, which is not the case, so we decided to use classes. You are free to mutate the elements without having to create copies all the time. That simplifies the developer's work a lot.

I'll merge this PR 🎉
